src/formatter.ts

```diff
@@ -0,0 +1,67 @@
+import * as cp from "child_process";
+import * as path from "path";
+import * as vscode from "vscode";
+
+export class Formatter implements vscode.DocumentFormattingEditProvider {
+    readonly binaryPath: string = '';
+
+    constructor(binaryPath: string) {
+        if (!binaryPath || binaryPath.length === 0) {
+            throw new Error('binaryPath is required to construct a formatter');
+        }
+        this.binaryPath = binaryPath;
+    }
+
+    public provideDocumentFormattingEdits(
+        document: vscode.TextDocument,
+        // @ts-ignore: The formatter doesn't have any options, but we need to implement the interface.
+        options: vscode.FormattingOptions,
+        token: vscode.CancellationToken
+    ): vscode.ProviderResult<vscode.TextEdit[]> {
+        if (vscode.window.visibleTextEditors.every((e) => e.document.fileName !== document.fileName)) {
+            return [];
+        }
+        return this.runFormatter(document, token).then(
+            (edits) => edits,
+            (err) => {
+                console.log(err);
+                return Promise.reject('Check the console in dev tools to find errors when formatting.');
+            }
+        );
+    }
+
+    private runFormatter(
+        document: vscode.TextDocument,
+        // @ts-ignore: TODO: Use the CancellationToken.
+        token: vscode.CancellationToken
+    ): Thenable<vscode.TextEdit[]> {
+        return new Promise<vscode.TextEdit[]>((resolve, reject) => {
+            const cwd = path.dirname(document.fileName);
+            let stdout = '';
+            let stderr = '';
+
+            // Use spawn instead of exec to avoid maxBufferExceeded error
+            const p = cp.spawn(this.binaryPath, [document.fileName], { cwd });
+            // TODO: We need to use the CancellationToken.
```

Comment:
See [this](https://github.com/golang/vscode-go/blob/0fa41b1c6e4217cc68a61bc97a7cabd34655e3cb/src/language/legacy/goFormat.ts#L87).